src/runtime_src/core/tools/common/tests/TestBandwidthKernel.cpp

```diff
@@ -25,8 +36,241 @@ TestBandwidthKernel::run(std::shared_ptr<xrt_core::device> dev)
+  const std::string test_path = findPlatformPath(dev, ptree);
+  const std::string b_file = findXclbinPath(dev, ptree); // bandwidth.xclbin
+  bool flag_s = false;
+
+  xrt::device device(dev->get_device_id());
+
+  if (test_path.empty()) {
+    logger(ptree, "Error", "Platform test path was not found.");
+    ptree.put("status", test_token_failed);
+    return;
+  }
+
+  auto retVal = validate_binary_file(b_file);
+  if (flag_s || retVal == EOPNOTSUPP) {
+    ptree.put("status", test_token_skipped);
+    return;
+  } else if (retVal != EXIT_SUCCESS) {
+    logger(ptree, "Error", "Unknown error validating bandwidth xclbin");
+    ptree.put("status", test_token_failed);
+    return;
+  }
+
+  int num_kernel = 0, num_kernel_ddr = 0;
+  bool chk_hbm_mem = false;
+  std::string filename = "platform.json";
+  auto platform_json = std::filesystem::path(test_path) / filename;
+
+  try {
+    boost::property_tree::ptree load_ptree_root;
+    boost::property_tree::read_json(platform_json.string(), load_ptree_root);
+
+    auto temp = load_ptree_root.get_child("total_ddr_banks");
+    num_kernel = temp.get_value<int>();
+    num_kernel_ddr = num_kernel;
+    auto pt_mem_array = load_ptree_root.get_child("meminfo");
+    for (const auto& mem_entry : pt_mem_array) {
+      boost::property_tree::ptree pt_mem_entry = mem_entry.second;
+      auto sValue = pt_mem_entry.get<std::string>("type");
+      if (sValue == "HBM")
+        chk_hbm_mem = true;
+    }
+    if (chk_hbm_mem) {
+      // As HBM is part of platform, number of ddr kernels is total count reduced by 1(single HBM)
+      num_kernel_ddr = num_kernel - 1;
+    }
+  } catch (const std::exception&) {
+    logger(ptree, "Error", "Bad JSON format detected while marshaling build metadata");
+    ptree.put("status", test_token_skipped);
+    return;
+  }
+
+  auto xclbin_uuid = device.load_xclbin(b_file);
+
+  std::string krnl_name = "bandwidth";
+  std::vector<xrt::kernel> krnls(num_kernel);
+
+  for (int i = 0; i < num_kernel; i++) {
+    std::string cu_id = std::to_string(i + 1);
+    std::string krnl_name_full = krnl_name + ":{" + "bandwidth_" + cu_id + "}";
+
+    // Here Kernel object is created by specifying kernel name along with
+    // compute unit.
+    // For such case, this kernel object can only access the specific
+    // Compute unit
+    krnls[i] = xrt::kernel(device, xclbin_uuid, krnl_name_full.c_str());
+  }
+
+  double max_throughput = 0;
+  int reps = 10000;
+  if (num_kernel_ddr) {
+    // Starting at 4K and going up to 16M with increments of power of 2
+    for (uint32_t a = 4 * 1024; a <= 16 * 1024 * 1024; a *= 2) {
+      unsigned int data_size = a;
+
+      if (is_emulation()) {
+        reps = 2; // reducing the repeat count to 2 for emulation flow
+        // Running only upto 8K for emulation flow
+        if (data_size > 8 * 1024) {
+          break;
+        }
+      }
+
+      unsigned int vector_size_bytes = data_size;
```

Comment:
there is too much happening in one func. Try to split it up in "init_buffer", "init_bo" etc. Look at this sample test case from Kelly: https://gitenterprise.xilinx.com/XRT/testcases/blob/master/xrt_qor_throughput/src/host_hal.cpp. This testcase is much more readable